File: src/com/interview/sde/algorithm/linkedlist/MergeTwoLinkedLists.java
package com.interview.sde.algorithm.linkedlist;

import java.util.Scanner;

//https://www.hackerrank.com/challenges/merge-two-sorted-linked-lists/problem
public class MergeTwoLinkedLists {

    private static final Scanner scanner = new Scanner(System.in);

    public static void printSinglyLinkedList(SinglyLinkedListNode node, String sep) {
        while (node != null) {
            System.out.println(node.data);

            node = node.next;

            if (node != null) {
                System.out.println(sep);
            }
        }
    }

    /*
     * For your reference:
     *
     * SinglyLinkedListNode {
     *     int data;
     *     SinglyLinkedListNode next;
     * }
     *
     */
    static SinglyLinkedListNode mergeLists(SinglyLinkedListNode head1, SinglyLinkedListNode head2) {
        if (head1 == null) {
            return head2;
        }
        if (head2 == null) {
            return head1;
        }
        SinglyLinkedListNode headNodeToReturn = (head1.data > head2.data) ? head2 : head1;

        while (head1 != null && head2 != null) {
            if (head1.data > head2.data) {
                head2 = walkOnList(head1, head2);

            } else {
                head1 = walkOnList(head2, head1);
            }
        }
        return headNodeToReturn;
    }

    // Complete the mergeLists function below.

    private static SinglyLinkedListNode walkOnList(SinglyLinkedListNode biggerNumberListHead, SinglyLinkedListNode smallerNumberListHead) {
        SinglyLinkedListNode holderHead;

        while (smallerNumberListHead.next != null && smallerNumberListHead.next.data <= biggerNumberListHead.data) {
            smallerNumberListHead = smallerNumberListHead.next;
        }

        holderHead = smallerNumberListHead.next;
        smallerNumberListHead.next = biggerNumberListHead;
        smallerNumberListHead = holderHead;
        return smallerNumberListHead;
    }

    //Implemented with mentee
    public static SinglyLinkedListNode mergeTwoLists(SinglyLinkedListNode l1, SinglyLinkedListNode l2) {

        SinglyLinkedListNode currentNode = null;
        SinglyLinkedListNode head = null;

        while (l1 != null || l2 != null) {
            if (l1 == null || l2 != null && l2.data < l1.data) {
                if (head == null) { //test if solution is empty
                    head = new SinglyLinkedListNode(l2.data);
                    currentNode = head;
                } else { // if it's not, create a new node, set the currentNode, walk the currentNode
                    currentNode.next = new SinglyLinkedListNode(l2.data);
                    currentNode = currentNode.next;
                }
                l2 = l2.next;
            } else {
                if (head == null) {
                    head = new SinglyLinkedListNode(l1.data);
                    currentNode = head;
                } else {
                    currentNode.next = new SinglyLinkedListNode(l1.data);
                    currentNode = currentNode.next;
                }
                l1 = l1.next;
            }
        }
        return head;
    }

    public static void main(String[] args) {

        int tests = scanner.nextInt();
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        for (int testsItr = 0; testsItr < tests; testsItr++) {
            SinglyLinkedList llist1 = new SinglyLinkedList();

            int llist1Count = scanner.nextInt();
            scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

            for (int i = 0; i < llist1Count; i++) {
                int llist1Item = scanner.nextInt();
                scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

                llist1.insertNode(llist1Item);
            }

            SinglyLinkedList llist2 = new SinglyLinkedList();

            int llist2Count = scanner.nextInt();
            scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

            for (int i = 0; i < llist2Count; i++) {
                int llist2Item = scanner.nextInt();
                scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

                llist2.insertNode(llist2Item);
            }

            SinglyLinkedListNode llist3 = mergeTwoLists(llist1.head, llist2.head);

            printSinglyLinkedList(llist3, " ");
            System.out.println();
        }


        scanner.close();
    }

    static class SinglyLinkedListNode {
        public int data;
        public SinglyLinkedListNode next;

        public SinglyLinkedListNode(int nodeData) {
            this.data = nodeData;
            this.next = null;
        }
    }

    static class SinglyLinkedList {
        public SinglyLinkedListNode head;
        public SinglyLinkedListNode tail;

        public SinglyLinkedList() {
            this.head = null;
            this.tail = null;
        }

        public void insertNode(int nodeData) {
            SinglyLinkedListNode node = new SinglyLinkedListNode(nodeData);

            if (this.head == null) {
                this.head = node;
            } else {
                this.tail.next = node;
            }

            this.tail = node;
        }
    }
}

File: src/com/interview/sde/java/linkedlist/MergeTwoLinkedLists.js
//https://www.hackerrank.com/challenges/merge-two-sorted-linked-lists/problem

'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

const SinglyLinkedListNode = class {
    constructor(nodeData) {
        this.data = nodeData;
        this.next = null;
    }
};

const SinglyLinkedList = class {
    constructor() {
        this.head = null;
        this.tail = null;
    }

    insertNode(nodeData) {
        const node = new SinglyLinkedListNode(nodeData);

        if (this.head == null) {
            this.head = node;
        } else {
            this.tail.next = node;
        }

        this.tail = node;
    }
};

function printSinglyLinkedList(node, sep, ws) {
    while (node != null) {
        ws.write(String(node.data));

        node = node.next;

        if (node != null) {
            ws.write(sep);
        }
    }
}

// Complete the mergeLists function below.

/*
 * For your reference:
 *
 * SinglyLinkedListNode {
 *     int data;
 *     SinglyLinkedListNode next;
 * }
 *
 */
function mergeLists(l1, l2) {

    let head = new SinglyLinkedListNode();

    if (l1) {
        if (!l2 || l2 && l1.data <= l2.data) {
            head = new SinglyLinkedListNode(l1.data);
            l1 = l1.next
        } else {
            head = new SinglyLinkedListNode(l2.data);
            l2 = l2.next
        }
    } else if (l2) {
        if (!l1 || l1 && l1.data <= l2.data) {
            head = new SinglyLinkedListNode(l1.data);
            l1 = l1.next
        } else {
            head = new SinglyLinkedListNode(l2.data);
            l2 = l2.next
        }
    }

    let current = head;

    while (l1 || l2) {

        if (l1) {
            if (!l2 || l2 && l1.data <= l2.data) {
                current.next = new SinglyLinkedListNode(l1.data);
                l1 = l1.next
                current = current.next
            } else {
                current.next = new SinglyLinkedListNode(l2.data);
                l2 = l2.next
                current = current.next
            }
        } else if (l2) {
            if (!l1 || l1 && l2.data <= l1.data) {
                current.next = new SinglyLinkedListNode(l2.data);
                l2 = l2.next
                current = current.next
            } else {
                current.next = new SinglyLinkedListNode(l1.data);
                l1 = l1.next
                current = current.next
            }
        }

    }
    return head;

}

function main() {
}
